plugins/incremental-ingestion-backend/src/database/IncrementalIngestionDatabaseManager.ts

```diff
@@ -0,0 +1,592 @@
+/*
+ * Copyright 2022 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Knex } from 'knex';
+import type { DeferredEntity } from '@backstage/plugin-catalog-backend';
+import { stringifyEntityRef } from '@backstage/catalog-model';
+import { Duration } from 'luxon';
+import { v4 } from 'uuid';
+import {
+  INCREMENTAL_ENTITY_PROVIDER_ANNOTATION,
+  IngestionRecord,
+  IngestionRecordUpdate,
+  IngestionUpsertIFace,
+  MarkRecord,
+  MarkRecordInsert,
+} from '../types';
+
+/** @public */
+export class IncrementalIngestionDatabaseManager {
```

Comment:
There are no tests for this class (might be good for at least specifically this one since it might be the riskier one in future refactors)

Reply:
I'll look into writing one. Is this a showstopper?